### PR TITLE
cpp: Add documentation about Pigweed as a solution

### PR DESCRIPTION
**Pull request by yperess**
Include documentation about:
- some C++ features via Pigweed
- instructions how to setup Pigweed

This PR addresses some aspects of #53851 by giving guidance but not making Pigweed the 'selected' approach to C++.

### CODE REVIEW

File: doc/develop/languages/cpp/pigweed.rst [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    you'll need to exit and delete it first. Pigweed will manage the virtual
+    environment for you. Trying to bootstrap from inside a virtual environment
+    will lead to undefined behavior.
+
+Bootstrapping the Pigweed environment is effectively like Zephyr's installation
+of the ``requirements.txt`` file. After you bootstrap Pigweed, there will be a
+virtual environment set up with all the Pigweed tools. Then, you install the
+Zephyr requirements on top.
+
+.. code-block:: console
```
Comment (kartben) [inline doc/develop/languages/cpp/pigweed.rst:85]:
```suggestion
.. code-block:: shell
```

File: doc/develop/languages/cpp/pigweed.rst [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+   cd ~/zephyrproject
+   source modules/lib/pigweed/bootstrap.sh
+   pip install zephyr/scripts/requirements.txt
+
+Once installed, the environment is set up. In future work sessions, run the
+following command to enter your virtual environment:
+
+.. code-block:: console
```
Comment (kartben) [inline doc/develop/languages/cpp/pigweed.rst:94]:
```suggestion
.. code-block:: shell
```

**Review by kartben — CHANGES_REQUESTED** [review]
this is great, thank you!

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
This goes against what was discussed in the TSC meeting on 2023-08-30 regarding recommending using pigweed to make up for currently unimplemented C++ features in the main Zephyr code (see MoM https://docs.google.com/document/d/1cMkXkFN-Yje-7zIGZmWMfkrVCju85UNnKDfjng9vEv0/edit#heading=h.8cm9xcy7g2ul).

Adding pigweed references to the "Language Support" section of the main Zephyr documentation leads users to believe that this is the approach recommended by the Zephyr project, which the TSC discussion did not conclude.

File: doc/develop/getting_started/index.rst [hunk 3]
```diff
@@ -178,6 +178,8 @@ The current minimum required version for the main dependencies are:
 .. _install_py_requirements:
 .. _gs_python_deps:
```
[the remark this answers is not in the document]
Comment (yperess) [reply doc/develop/getting_started/index.rst:180]: Done

File: doc/develop/languages/cpp/pigweed.rst [hunk 4]
```diff
@@ -0,0 +1,115 @@
+.. _pigweed_cpp_support:
+
+Enable C++ Support with Pigweed
```
[the remark this answers is not in the document]
Comment (yperess) [reply doc/develop/languages/cpp/pigweed.rst:3]: Done

File: doc/develop/languages/cpp/pigweed.rst (hunk 1 above)
Comment (yperess) [reply doc/develop/languages/cpp/pigweed.rst:85]: done

**Comment by yperess** [comment]
> This goes against what was discussed in the TSC meeting on 2023-08-30 regarding recommending using pigweed to make up for currently unimplemented C++ features in the main Zephyr code (see MoM https://docs.google.com/document/d/1cMkXkFN-Yje-7zIGZmWMfkrVCju85UNnKDfjng9vEv0/edit#heading=h.8cm9xcy7g2ul).
> 
> Adding pigweed references to the "Language Support" section of the main Zephyr documentation leads users to believe that this is the approach recommended by the Zephyr project, which the TSC discussion did not conclude.

I read through all the notes before putting this PR together. I'm happy to discuss further, but we never excluded this from going into documentation. The discussion went around the use of it as a module and the delegation of saying "we're going to make this official". Unfortunately the meetings notes seem to be missing a key detail that I'm recalling which is that we did discuss having documentation for modules that we want to advertise but aren't in the manifest yet. Putting this in the C++ documentation is the most natural place IMO.

Happy to discuss further.

**Comment by carlescufi** [comment]
@henrikbrixandersen @yperess @nashif 
I think a middle-ground may work here. We could add documentation for Pigweed in a separate section (i.e. move `pigweed.rst` elsewhere which by the way in my opinion is separate from the goal pursued in https://github.com/zephyrproject-rtos/zephyr/pull/61505, which just documents a list of all modules, and does not add specific module documentation) and then link to it from the C++ language support page:

`for an alterntative approach to C++ support please see Pigweed`

But for that we would need  `modules` top-level section in the doc that could document at least the default modules provided by Zephyr.

**Comment by nashif** [comment]
> goal pursued in #61505, which just documents a list of all modules, and does not add specific module documentation) and then link to it from the C++ language support page:

that is not correct, see https://github.com/zephyrproject-rtos/zephyr/pull/61505/files#diff-c22a8fff0128e724a948e334206112f304cb2c0a00e4429abade35390dd7d890R130

**Comment by yperess** [comment]
> @yperess  the idea was to add such documentation under external modules as being introduced in this PR https://github.com/zephyrproject-rtos/zephyr/pull/61505
> 
> 
> 

Sure, seems easy enough to just relocate the pigweed.rst. I just put it where people would most naturally look for it. We can either:

- merge this and move it when your PR lands
- wait and rebase when your PR lands

Personally I would prefer to land and move later just to have less things up in the air at the same time. The current layout (without the optional modules) makes sense to me and even with optional modules I would put a `:ref:` to the Pigweed docs from the C++ docs. So i guess I'm not seeing the harm in this being an interim solution.

**Comment by carlescufi** [comment]
> > goal pursued in #61505, which just documents a list of all modules, and does not add specific module documentation) and then link to it from the C++ language support page:
> 
> that is not correct, see https://github.com/zephyrproject-rtos/zephyr/pull/61505/files#diff-c22a8fff0128e724a948e334206112f304cb2c0a00e4429abade35390dd7d890R130

OK, sorry I misunderstood this. Then I suggest we merge #61505 first and then add the link to it from the C++ language page.

**Comment by kartben** [comment]
> OK, sorry I misunderstood this. Then I suggest we merge #61505 first and then add the link to it from the C++ language page.

fwiw #61605 has now been merged. cc @yperess

**Comment by kartben** [comment]
@yperess did you want to revisit this one?

**Comment by yperess** [comment]
> @yperess did you want to revisit this one?

Yes please, I'll rebase

**Comment by yperess** [comment]
No code/doc changes, just a raw rebase. @kartben @nashif , what's the process for this now? Do I need to propose Pigweed as an addition to `submanifests/optional.yaml`? The original goal here was just to document that the option exists, I worry that adding it to the submanifest and having to request a fork will be a bit too much right now.

**Comment by kartben** [comment]
@nashif can you help here? I don't think we have a requirement for *optional* modules to live as a fork under the `zephyrproject-rtos` GH org but I am not sure.

**Comment by fabiobaltieri** [comment]
> @nashif can you help here? I don't think we have a requirement for _optional_ modules to live as a fork under the `zephyrproject-rtos` GH org but I am not sure.

Well they are fetched in CI, worth having the conversation but it may not be a bad idea to keep doing so.

**Comment by nashif** [comment]
> @nashif can you help here? I don't think we have a requirement for _optional_ modules to live as a fork under the `zephyrproject-rtos` GH org but I am not sure.

there is a requirement, ?"optional modules" are not enabled/active by default, but otherwise follow the same requirement as enabled and active modules, external projects are those which do not exist in the manifest and do not need to be in the zephyr project org:

https://docs.zephyrproject.org/latest/contribute/external.html#integration-as-external-modules

**Comment by nashif** [comment]
just submitted a PR for the template... 

#66860

**Pull request closed without merging**